ikarus/finiteelements/mechanics/materials/hyperelastic/interface.hh

```diff
@@ -0,0 +1,348 @@
+
+      auto realDev = dev_.template rebind<double>();
+      e.val        = realDev.storedEnergy(lambdas);
+
+      e.grad.val         = (transformDeviatoricStresses(realDev.stresses(lambdas), N).transpose() / 2 * dualC).trace();
+      e.val.grad         = e.grad.val;
+      const auto Cmoduli = transformDeviatoricTangentModuli(realDev.tangentModuli(lambdas), N);
+
+      Eigen::array<Eigen::IndexPair<Eigen::Index>, 2> double_contraction  = {Eigen::IndexPair<Eigen::Index>(2, 0),
+                                                                             Eigen::IndexPair<Eigen::Index>(3, 1)};
+      Eigen::array<Eigen::IndexPair<Eigen::Index>, 2> double_contraction2 = {Eigen::IndexPair<Eigen::Index>(0, 0),
+                                                                             Eigen::IndexPair<Eigen::Index>(1, 1)};
+      const auto tCdual                  = tensorView(dualC, std::array<Eigen::Index, 2>({3, 3}));
+      const auto tCdualT                 = tensorView(dualC2, std::array<Eigen::Index, 2>({3, 3}));
+      const auto prod                    = Cmoduli.contract(tCdual, double_contraction);
+      const Eigen::Tensor<double, 0> res = tCdualT.contract(prod, double_contraction2);
+      e.grad.grad                        = res(0) / 4.0; // extracting value of zero order tensor
+
+      return e;
+    } else
+      static_assert(Dune::AlwaysFalse<Derived>::value, "No fitting ScalarType.");
+  }
+
+  /** \brief A helper function to compute the deviatoric part of the stresses.
+   *
+   * \details While using AutoDiff, if the eigenvalues (principal stretches) of C are degenerated, the derivative can
+   * have certain singularities. This is circumvented here by explicitly updating the derivatives.
+   *
+   */
+  template <typename Derived>
+  auto deviatoricStress(const Eigen::MatrixBase<Derived>& C) const {
+    if constexpr (not Concepts::AutodiffScalar<typename Derived::Scalar>) {
+      auto [lambdas, N] = principalStretches(C);
+      return transformDeviatoricStresses(dev_.stresses(lambdas), N);
+    } else if constexpr (std::is_same_v<ScalarType, autodiff::dual>) {
+      constexpr int nVoigtIndices = 6;
+      Eigen::Vector<autodiff::dual, nVoigtIndices> g;
+      auto Cvec           = toVoigt(C.derived());
+      const auto realCVec = derivative<0>(Cvec);
+      auto realC          = fromVoigt(realCVec);
+      auto dualC          = fromVoigt(forEach(Cvec, [](const auto& v) { return v.grad; }).eval());
+      auto [lambdas, N]   = principalStretches(realC);
+
+      auto realDev       = dev_.template rebind<double>();
+      const auto Cmoduli = toVoigt(transformDeviatoricTangentModuli(realDev.tangentModuli(lambdas), N));
+      for (int i = 0; i < nVoigtIndices; ++i) {
+        Eigen::Vector<double, nVoigtIndices> contraction = Cmoduli * toVoigt(dualC);
+        contraction.topRows<3>() /= 2.0;
+        g[i].val  = toVoigt(transformDeviatoricStresses(realDev.stresses(lambdas), N))[i];
```

Comment:
This recomputes everything in each loop. Pleas compute the stresses and their transformations outside the loop.